File: 2015/21/python/main.py
from pathlib import Path
from typing import Dict, Tuple, List
from itertools import combinations

class Entity:
    def __init__(self, hp: int, armor: int, dmg: int):
        self.hp = hp
        self.armor = armor
        self.dmg = dmg
    
    def is_alive(self) -> bool:
        return self.hp > 0
    
    def take_damage(self, incoming_damage: int) -> None:
        actual_damage = max(1, incoming_damage - self.armor)
        self.hp -= actual_damage
    
    def attack(self, target: 'Entity') -> None:
        target.take_damage(self.dmg)

def simulate_combat(player_stats: Dict[str, int]) -> bool:
    
    player = Entity(
        hp=100,  
        armor=player_stats['Armor'],
        dmg=player_stats['Damage']
    )
    
    boss = Entity(
        hp=109,  
        armor=2,
        dmg=8
    )
    
    while True:
        player.attack(boss)
        if not boss.is_alive():
            return True  # Player wins
            
        boss.attack(player)
        if not player.is_alive():
            return False  # Boss wins

def parse_items(path: str) -> Tuple[Dict[str, Dict[str, int]], Dict[str, Dict[str, int]], Dict[str, Dict[str, int]]]:
    data = Path("../items.txt").read_text().strip().split("\n\n")
    weapons = {}
    armors = {}
    rings = {}
    for i, cat in enumerate(data):
        if i == 0:
            data = cat.splitlines()
            for itm in data[1:]:
                info = itm.split()
                if info[0].strip() not in weapons: weapons[info[0].strip()] = {}
                weapons[info[0].strip()]["Cost"] = int(info[1].strip())
                weapons[info[0].strip()]["Damage"] = int(info[2].strip())
                weapons[info[0].strip()]["Armor"] = int(info[-1].strip())
        elif i == 1:
            data = cat.splitlines()
            for itm in data[1:]:
                info = itm.split()
                if info[0].strip() not in armors: armors[info[0].strip()] = {}
                armors[info[0].strip()]["Cost"] = int(info[1].strip())
                armors[info[0].strip()]["Damage"] = int(info[2].strip())
                armors[info[0].strip()]["Armor"] = int(info[-1].strip())
        elif i == 2:
            data = cat.splitlines()
            for itm in data[1:]:
                info = itm.split("   ")
                if info[0].strip() not in rings: rings[info[0].strip()] = {}
                rings[info[0].strip()]["Cost"] = int(info[1].strip())
                rings[info[0].strip()]["Damage"] = int(info[2].strip())
                rings[info[0].strip()]["Armor"] = int(info[-1].strip())
    return weapons, armors, rings

def gen_item_combinations(weapons: Dict[str, Dict[str, int]], 
                        armors: Dict[str, Dict[str, int]], 
                        rings: Dict[str, Dict[str, int]]) -> List[Tuple[str, ...]]:
    valid_combinations = []
    
    # Generate ring combinations (0, 1, or 2 rings)
    ring_combinations = [()]
    ring_combinations.extend(combinations(rings.keys(), 1))  # One ring
    ring_combinations.extend(combinations(rings.keys(), 2))  # Two rings
    
    # Generate armor combinations (0 or 1 armor)
    armor_combinations = [()]
    armor_combinations.extend((armor_name,) for armor_name in armors.keys())
    
    for weapon_name in weapons.keys():
        for armor_combo in armor_combinations:
            for ring_combo in ring_combinations:
                combination = (weapon_name,) + armor_combo + ring_combo
                valid_combinations.append(combination)
    
    return valid_combinations

def calculate_stats(combination: Tuple[str, ...], 
                   weapons: Dict[str, Dict[str, int]], 
                   armors: Dict[str, Dict[str, int]], 
                   rings: Dict[str, Dict[str, int]]) -> Dict[str, int]:
    total_stats = {'Cost': 0, 'Damage': 0, 'Armor': 0}
    
    for item in combination:
        if item in weapons:
            stats = weapons[item]
        elif item in armors:
            stats = armors[item]
        elif item in rings:
            stats = rings[item]
        else:
            continue
            
        for stat in total_stats:
            total_stats[stat] += stats[stat]
            
    return total_stats

def find_optimal_equipment(weapons: Dict[str, Dict[str, int]],
                         armors: Dict[str, Dict[str, int]],
                         rings: Dict[str, Dict[str, int]],
                         optimize_for_cost: bool = True) -> Tuple[Tuple[str, ...], Dict[str, int]]:
    combinations = gen_item_combinations(weapons, armors, rings)
    optimal_combo = None
    optimal_stats = None
    optimal_cost = float('inf') if optimize_for_cost else float('-inf')
    
    for combo in combinations:
        stats = calculate_stats(combo, weapons, armors, rings)
        player_wins = simulate_combat(stats)
        
        if optimize_for_cost:
            if player_wins and stats['Cost'] < optimal_cost:
                optimal_combo = combo
                optimal_stats = stats
                optimal_cost = stats['Cost']
        else:
            if not player_wins and stats['Cost'] > optimal_cost:
                optimal_combo = combo
                optimal_stats = stats
                optimal_cost = stats['Cost']
    
    return optimal_combo, optimal_stats

def main() -> None:
    item_path = "../items.txt"
    weapons, armors, rings = parse_items(item_path)
    
    winning_combo, winning_stats = find_optimal_equipment(weapons, armors, rings, optimize_for_cost=True)
    print("=" * 20, "Part 1", "=" * 20)
    print("Cheapest winning combination:")
    print(f"Items: {winning_combo}")
    print(f"Stats: {winning_stats}")
    print(f"Total cost: {winning_stats['Cost']} gold\n")

    losing_combo, losing_stats = find_optimal_equipment(weapons, armors, rings, optimize_for_cost=False)
    print("=" * 20, "Part 2", "=" * 20)
    print("Most expensive losing combination:")
    print(f"Items: {losing_combo}")
    print(f"Stats: {losing_stats}")
    print(f"Total cost: {losing_stats['Cost']} gold")

if __name__ == "__main__":
    main()

File: 2015/21/rust/src/main.rs
use std::{collections::HashMap, fs};
use itertools::Itertools;

#[derive(Copy,Clone)]
struct Entity {
    hp: i32,
    armor: i32,
    dmg: i32,
}

impl Entity {
    pub fn new(hp: i32, armor: i32, dmg: i32) -> Entity {
        Entity {hp, armor, dmg}
    }
    pub fn is_alive(self) -> bool {
        self.hp > 0 
    }
    pub fn take_damage(&mut self, dmg: i32) {
        let damage = dmg - self.armor;
        let actual_damage = damage.max(1);
        self.hp -= actual_damage;
    }
    pub fn attack(self, target: &mut Entity) {
        target.take_damage(self.dmg);
    }
}

fn simulate_combat(player_stats: HashMap<String, i32>) -> bool {
    let mut player = Entity::new(
        100,
        *player_stats.get("Armor").unwrap_or(&0), 
        *player_stats.get("Damage").unwrap_or(&0));
    let mut boss = Entity::new(109, 2, 8);
    
    loop {
        player.attack(&mut boss);
        if !boss.is_alive(){
            return true;
        }

        boss.attack(&mut player);
        if !player.is_alive(){
            return false
        }
    }
}

fn parse_items(path: &str) ->
(HashMap<String, HashMap<String, i32>>, 
HashMap<String, HashMap<String, i32>>, 
HashMap<String, HashMap<String, i32>>) {
    let binding = fs::read_to_string(path)
    .expect("Error parsing input");
    let data: Vec<&str> = binding
    .trim()
    .split("\n\n")
    .collect();

    let mut weapons: HashMap<String, HashMap<String, i32>> = HashMap::new();
    let mut armors: HashMap<String, HashMap<String, i32>> = HashMap::new();
    let mut rings: HashMap<String, HashMap<String, i32>> = HashMap::new();

    for (i,cat) in data.iter().enumerate(){
        let lines: Vec<&str> = cat.lines().collect();

        for item in &lines[1..]{
            let info: Vec<&str> = if i == 2 {
                item.split("   ").collect()
            } else {
                item.split_whitespace().collect()
            };

            if info.len() < 4 {continue;}

            let name = info[0].trim().to_string();
            let cost = info[1].trim().parse::<i32>().unwrap_or(0);
            let damage = info[2].trim().parse::<i32>().unwrap_or(0);
            let armor = info[3].trim().parse::<i32>().unwrap_or(0);

            let entry = HashMap::from([
                ("Cost".to_string(), cost),
                ("Damage".to_string(), damage),
                ("Armor".to_string(), armor),
            ]);

            match i {
                0 => {weapons.insert(name, entry);}
                1 => {armors.insert(name, entry);}
                2 => {rings.insert(name, entry);}
                _ => {}
            }
        }
    }
    (weapons, armors, rings)
}

fn gen_item_combinations(
    weapons: &HashMap<String, HashMap<String, i32>>,
    armors: &HashMap<String, HashMap<String, i32>>,
    rings: &HashMap<String, HashMap<String, i32>>,
) -> Vec<Vec<String>> {
    
    let mut valid_combinations = Vec::new();

    // Generate ring combinations (0, 1, or 2 rings)
    let ring_combinations: Vec<Vec<String>> = vec![vec![]]
        .into_iter()
        .chain(rings.keys().map(|r| vec![r.clone()]))
        .chain(
            rings.keys()
                .combinations(2)
                .map(|r| r.into_iter().cloned().collect())
        )
        .collect();

    // Generate armor combinations (0 or 1 armor)
    let armor_combinations: Vec<Vec<String>> = vec![vec![]]
        .into_iter()
        .chain(armors.keys().map(|a| vec![a.clone()]))
        .collect();

    // Generate all valid combinations
    for weapon in weapons.keys() {
        for armor in &armor_combinations {
            for ring in &ring_combinations {
                let mut comb = vec![weapon.clone()];
                comb.extend_from_slice(armor);
                comb.extend_from_slice(ring);
                valid_combinations.push(comb);
            }
        }
    }

    valid_combinations
}

fn calculate_stats(
    combination: &[String],
    weapons: &HashMap<String, HashMap<String, i32>>, 
    armors: &HashMap<String, HashMap<String, i32>>, 
    rings: &HashMap<String, HashMap<String, i32>>,
) -> HashMap<String, i32> {
    let mut total = HashMap::from([
        ("Cost".to_string(), 0),
        ("Damage".to_string(), 0),
        ("Armor".to_string(), 0),
    ]);
    
    for item in combination {
        let stats = weapons.get(item).or_else(|| armors.get(item)).or_else(|| rings.get(item));
        
        if let Some(stats) = stats {
            for (stat, value) in stats {
                if let Some(entry) = total.get_mut(stat) {
                    *entry += value;
                }
            }
        }
    }
    
    total
}


fn find_optimal_equipment(
weapons: &HashMap<String, HashMap<String, i32>>,
armors: &HashMap<String, HashMap<String, i32>>,
rings: &HashMap<String, HashMap<String, i32>>,
optimize_for_cost: bool,
) -> (Vec<String>, HashMap<String, i32>){
    let mut optimal_comb = Vec::new();
    let mut optimal_stats = HashMap::new();
    let mut optimal_cost = if optimize_for_cost {i32::MAX} else {i32::MIN};

    for combo in gen_item_combinations(weapons, armors, rings){
        let stats = calculate_stats(&combo, weapons, armors, rings);
        let player_wins = simulate_combat(stats.clone());

        if (optimize_for_cost && player_wins && stats["Cost"] < optimal_cost)
        || (!optimize_for_cost && !player_wins && stats["Cost"] > optimal_cost) 
        {
            optimal_cost = stats["Cost"];
            optimal_comb = combo.clone();
            optimal_stats = stats;
        }
    }
    (optimal_comb, optimal_stats)
}

fn main() {
    let sep = "=".repeat(20);
    let (weapons, armors, rings) = parse_items("../items.txt");
    
    let (winning_combo, winning_stats) = find_optimal_equipment(&weapons, &armors, &rings, true);
    
    println!("{} Part 1 {}", sep, sep);
    println!("Cheapest winning combination: {:?}", winning_combo);
    println!("Stats: {:?}", winning_stats);
    println!("Total cost: {} gold\n", winning_stats["Cost"]);
    
    println!("{} Part 2 {}", sep, sep);
    let (losing_combo, losing_stats) = find_optimal_equipment(&weapons, &armors, &rings, false);
    println!("Most expensive losing combination: {:?}", losing_combo);
    println!("Stats: {:?}", losing_stats);
    println!("Total cost: {} gold", losing_stats["Cost"]);
}
